### PR TITLE
OpenMPI: Prefer Last 3.X

### PR DESCRIPTION
**Pull request by ax3l**
The latest OpenMPI release, v4.0.0, does not build with many GCC/GFortran variants. Since `+mpi` and `+fortran` are our defaults, a lot of users get hit.

Let's wait for some point releases. cc @jrood-nrel 

References:
- #9888
- #9917
- https://github.com/open-mpi/ompi/issues/6105
- https://github.com/open-mpi/ompi/issues/6106

### CODE REVIEW

**Comment by ggouaillardet** [comment]
@davydden some removed/deprecated symbols/subroutines are no more available by default, and this is likely the root cause of the `hdf5` situation.
Ideally, all packages would stop using these symbols.
Meanwhile, you can `configure --enable-mpi1-compatibility ...` to have these symbols/subroutines part of Open MPI 4.
IIRC, there is a bug since these symbols/subroutines should be in the Open MPI libraries but __not__ in the header files

The fortran is issue will be very likely fixed in Open MPI 4.0.1.
Since you are building `hdf5` libraries and friends from sources, you will need the `--enable-mpi1-compatibility`  configure option (and it is very unlikely we will make it the default in 4.0.1).

Bottom line, I believe you need the `--enable-mpi1-compatibility` configure option from Open MPI 4, and that being said, it is up to you to wait for 4.0.1 or start 4.0.0 today with the `--enable-mpi-ext=affinity,cuda` workaround.

**Comment by ax3l** [comment]
@ggouaillardet thank you for the details!
A quick question on [extensions in OpenMPI](https://github.com/open-mpi/ompi/wiki/MPIExtensions): are `affinity`, `cuda`, and Persistent collective (keyword?) the only extensions? Can you link a full list and what is default on?

Should one only build the `cuda` extension when CUDA is truly available or can it always be built? E.g. `--with-cuda=...`/`--without-cuda` variant.

Can other extensions pull further dependencies?

**Comment by ggouaillardet** [comment]
From Open MPI 4, the extensions are `affinity,cr,cuda,pcollreq` and the default is to __try__ to build them all. IIRC, `cr` requires some extra dependency and `configure` should (I did not test) abort if it is requested but cannot be built.

The `cuda` extension simply queries at runtime whether CUDA hardware is available.
Since you `configure —without-cuda`, the runtime will simply return a hard coded “no”.
Back to your question, you can always build this extension.

IIRC, `affinity` only requires `hwloc` which is already an Open MPI pre-requisite.

`pcollreq` has no dependency.

**Comment by jrood-nrel** [comment]
I agree 3.1.3 should be the default. I should have set that originally. I also just recently had issues with HDF5 and realized I should have done so.

**Comment by adamjstewart** [comment]
Pinged @tgamblin and it sounds like he is in agreement with this PR. Merging...

**Comment by rblake-llnl** [comment]
Does anyone here have a reproducer so we can tell if/when openmpi 4.0.1 has fixed the problems?

**Comment by ax3l** [comment]
In #11029, we want to add `--enable-mpi1-compatibility` as well. A Fortran build issue with Spack was also [fixed upstream in 4.0.1](https://raw.githubusercontent.com/open-mpi/ompi/v4.0.x/NEWS).

You could try to build apps and libs that depend on hdf5 with various GCC variants, such as 4.9 which is still commonly used. I personally tried the spack `adios`, `adios2`, `libsplash`, `openpmd-api` package, etc. with default variants (some are C++ packages, but spack default variants build with Fortran bindings on).

Please contribute your findings to the comments in #11029.
